# Assignment 1
## GHCN Data Analysis using Spark

# Task 2: Analysis

In [2]:
# Run this cell to import pyspark and to define start_spark() and stop_spark()

import findspark

findspark.init()

import getpass
import pandas as pd
import pyspark
import random
import re

from IPython.display import display, HTML
from pyspark import SparkContext
from pyspark.sql import SparkSession


# Constants used to interact with Azure Blob Storage using the hdfs command or Spark

global username

username = re.sub('@.*', '', getpass.getuser())

global azure_account_name
global azure_data_container_name
global azure_user_container_name
global azure_user_token

azure_account_name = "madsstorage002"
azure_data_container_name = "campus-data"
azure_user_container_name = "campus-user"
azure_user_token = r"sp=racwdl&st=2024-09-19T08:03:31Z&se=2025-09-19T16:03:31Z&spr=https&sv=2022-11-02&sr=c&sig=kMP%2BsBsRzdVVR8rrg%2BNbDhkRBNs6Q98kYY695XMRFDU%3D"


# Functions used below

def dict_to_html(d):
    """Convert a Python dictionary into a two column table for display.
    """

    html = []

    html.append(f'<table width="100%" style="width:100%; font-family: monospace;">')
    for k, v in d.items():
        html.append(f'<tr><td style="text-align:left;">{k}</td><td>{v}</td></tr>')
    html.append(f'</table>')

    return ''.join(html)


def show_as_html(df, n=20):
    """Leverage existing pandas jupyter integration to show a spark dataframe as html.
    
    Args:
        n (int): number of rows to show (default: 20)
    """

    display(df.limit(n).toPandas())

    
def display_spark():
    """Display the status of the active Spark session if one is currently running.
    """
    
    if 'spark' in globals() and 'sc' in globals():

        name = sc.getConf().get("spark.app.name")

        html = [
            f'<p><b>Spark</b></p>',
            f'<p>The spark session is <b><span style="color:green">active</span></b>, look for <code>{name}</code> under the running applications section in the Spark UI.</p>',
            f'<ul>',
            f'<li><a href="http://localhost:{sc.uiWebUrl.split(":")[-1]}" target="_blank">Spark Application UI</a></li>',
            f'</ul>',
            f'<p><b>Config</b></p>',
            dict_to_html(dict(sc.getConf().getAll())),
            f'<p><b>Notes</b></p>',
            f'<ul>',
            f'<li>The spark session <code>spark</code> and spark context <code>sc</code> global variables have been defined by <code>start_spark()</code>.</li>',
            f'<li>Please run <code>stop_spark()</code> before closing the notebook or restarting the kernel or kill <code>{name}</code> by hand using the link in the Spark UI.</li>',
            f'</ul>',
        ]
        display(HTML(''.join(html)))
        
    else:
        
        html = [
            f'<p><b>Spark</b></p>',
            f'<p>The spark session is <b><span style="color:red">stopped</span></b>, confirm that <code>{username} (notebook)</code> is under the completed applications section in the Spark UI.</p>',
            f'<ul>',
            f'<li><a href="http://mathmadslinux2p.canterbury.ac.nz:8080/" target="_blank">Spark UI</a></li>',
            f'</ul>',
        ]
        display(HTML(''.join(html)))


# Functions to start and stop spark

def start_spark(executor_instances=2, executor_cores=1, worker_memory=1, master_memory=1):
    """Start a new Spark session and define globals for SparkSession (spark) and SparkContext (sc).
    
    Args:
        executor_instances (int): number of executors (default: 2)
        executor_cores (int): number of cores per executor (default: 1)
        worker_memory (float): worker memory (default: 1)
        master_memory (float): master memory (default: 1)
    """

    global spark
    global sc

    cores = executor_instances * executor_cores
    partitions = cores * 4
    port = 4000 + random.randint(1, 999)

    spark = (
        SparkSession.builder
        .config("spark.driver.extraJavaOptions", f"-Dderby.system.home=/tmp/{username}/spark/")
        .config("spark.dynamicAllocation.enabled", "false")
        .config("spark.executor.instances", str(executor_instances))
        .config("spark.executor.cores", str(executor_cores))
        .config("spark.cores.max", str(cores))
        .config("spark.driver.memory", f'{master_memory}g')
        .config("spark.executor.memory", f'{worker_memory}g')
        .config("spark.driver.maxResultSize", "0")
        .config("spark.sql.shuffle.partitions", str(partitions))
        .config("spark.kubernetes.container.image", "madsregistry001.azurecr.io/hadoop-spark:v3.3.5-openjdk-8")
        .config("spark.kubernetes.container.image.pullPolicy", "IfNotPresent")
        .config("spark.kubernetes.memoryOverheadFactor", "0.3")
        .config("spark.memory.fraction", "0.1")
        .config(f"fs.azure.sas.{azure_user_container_name}.{azure_account_name}.blob.core.windows.net",  azure_user_token)
        .config("spark.app.name", f"{username} (notebook)")
        .getOrCreate()
    )
    sc = SparkContext.getOrCreate()
    
    display_spark()

    
def stop_spark():
    """Stop the active Spark session and delete globals for SparkSession (spark) and SparkContext (sc).
    """

    global spark
    global sc

    if 'spark' in globals() and 'sc' in globals():

        spark.stop()

        del spark
        del sc

    display_spark()


# Make css changes to improve spark output readability

html = [
    '<style>',
    'pre { white-space: pre !important; }',
    'table.dataframe td { white-space: nowrap !important; }',
    'table.dataframe thead th:first-child, table.dataframe tbody th { display: none; }',
    '</style>',
]
display(HTML(''.join(html)))

In [3]:

# Run this cell to start a spark session in this notebook
start_spark(executor_instances=6, executor_cores=2, worker_memory=6, master_memory=6)


25/04/09 12:28:58 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


spark.dynamicAllocation.enabled,false
spark.fs.azure.sas.uco-user.madsstorage002.blob.core.windows.net,"""sp=racwdl&st=2024-09-19T08:00:18Z&se=2025-09-19T16:00:18Z&spr=https&sv=2022-11-02&sr=c&sig=qtg6fCdoFz6k3EJLw7dA8D3D8wN0neAYw8yG4z4Lw2o%3D"""
spark.kubernetes.driver.pod.name,spark-master-driver
spark.sql.shuffle.partitions,48
spark.fs.azure.sas.campus-user.madsstorage002.blob.core.windows.net,"""sp=racwdl&st=2024-09-19T08:03:31Z&se=2025-09-19T16:03:31Z&spr=https&sv=2022-11-02&sr=c&sig=kMP%2BsBsRzdVVR8rrg%2BNbDhkRBNs6Q98kYY695XMRFDU%3D"""
spark.kubernetes.container.image.pullPolicy,IfNotPresent
spark.app.id,spark-60720b6c567d45b29adad8558f850e0c
spark.executor.instances,6
spark.app.startTime,1744158538560
spark.driver.memory,6g
fs.azure.sas.campus-user.madsstorage002.blob.core.windows.net,sp=racwdl&st=2024-09-19T08:03:31Z&se=2025-09-19T16:03:31Z&spr=https&sv=2022-11-02&sr=c&sig=kMP%2BsBsRzdVVR8rrg%2BNbDhkRBNs6Q98kYY695XMRFDU%3D


In [6]:

import math
import os

In [8]:

from pyspark.sql.types import *
from pyspark.sql.functions import to_date, col, count, when, collect_list
from pyspark.sql.functions import substring, trim
from pyspark.sql.functions import min, max, countDistinct
from pyspark.sql.functions import udf
from pyspark.sql.types import DoubleType
from pyspark.sql.functions import slice
from pyspark.sql.functions import collect_set, array_contains


# Question-1

In [9]:

# Define the path to save Parquet file
enriched_path = "wasbs://campus-user@madsstorage002.blob.core.windows.net/asa349/enriched_stations.parquet"

# Read the Parquet file into a new DataFrame
enriched_stations = spark.read.parquet(enriched_path)

# Show the first few rows
enriched_stations.show(5)


+-----------+----------+------------+--------+---------+---------+--------------------+--------+------------+------+--------------------+----------+---------+--------+--------------+-------------+--------------+
|         ID|STATE_CODE|COUNTRY_CODE|LATITUDE|LONGITUDE|ELEVATION|                NAME|GSN_FLAG|HCN_CRN_FLAG|WMO_ID|        COUNTRY_NAME|STATE_NAME|FIRSTYEAR|LASTYEAR|TOTAL_ELEMENTS|CORE_ELEMENTS|OTHER_ELEMENTS|
+-----------+----------+------------+--------+---------+---------+--------------------+--------+------------+------+--------------------+----------+---------+--------+--------------+-------------+--------------+
|AE000041196|      NULL|          AE|  25.333|   55.517|     34.0|SHARJAH INTER. AI...|     GSN|        NULL| 41196|United Arab Emira...|      NULL|     1944|    2025|             4|            3|             1|
|AEM00041218|      NULL|          AE|  24.262|   55.609|    264.9|AL AIN INTL      ...|    NULL|        NULL| 41218|United Arab Emira...|      NULL|    

## (a.1) How many stations are there in total?

In [10]:

enriched_stations.printSchema()


root
 |-- ID: string (nullable = true)
 |-- STATE_CODE: string (nullable = true)
 |-- COUNTRY_CODE: string (nullable = true)
 |-- LATITUDE: float (nullable = true)
 |-- LONGITUDE: float (nullable = true)
 |-- ELEVATION: float (nullable = true)
 |-- NAME: string (nullable = true)
 |-- GSN_FLAG: string (nullable = true)
 |-- HCN_CRN_FLAG: string (nullable = true)
 |-- WMO_ID: string (nullable = true)
 |-- COUNTRY_NAME: string (nullable = true)
 |-- STATE_NAME: string (nullable = true)
 |-- FIRSTYEAR: integer (nullable = true)
 |-- LASTYEAR: integer (nullable = true)
 |-- TOTAL_ELEMENTS: long (nullable = true)
 |-- CORE_ELEMENTS: long (nullable = true)
 |-- OTHER_ELEMENTS: long (nullable = true)



In [11]:

total_stations = enriched_stations.select("ID").distinct().count()
print(f"Total number of stations: {total_stations}")


[Stage 2:=======================================>                   (6 + 3) / 9]

Total number of stations: 129657


## (a.2) How many stations were active so far in 2025?

In [12]:

# Count stations that were active in 2024
active_stations_2025 = enriched_stations.filter(col("LASTYEAR") == 2025).count()
print(f"Number of stations active 2025 (Station Dataset): {active_stations_2025}")


Number of stations active 2025 (Station Dataset): 30735


In [13]:

# Define schema for daily datset
daily_schema = StructType([
    StructField("ID", StringType(), False),         # Station Code
    StructField("DATE", StringType(), False),       # Read as String initially (YYYYMMDD)
    StructField("ELEMENT", StringType(), False),    # Element Type Indicator
    StructField("VALUE", FloatType(), False),       # Data value
    StructField("MEASUREMENT_FLAG", StringType(), True),  # Measurement Flag
    StructField("QUALITY_FLAG", StringType(), True),      # Quality Flag
    StructField("SOURCE_FLAG", StringType(), True),       # Source Flag
    StructField("OBSERVATION_TIME", StringType(), True)   # Read as String first (HHMM)
])


In [14]:

# Define both input paths
daily_2025_path = f'wasbs://{azure_data_container_name}@{azure_account_name}.blob.core.windows.net/ghcnd/daily/2025.csv.gz'


In [15]:

# Read both CSVs with the same schema
daily_2025_df = spark.read.csv(daily_2025_path, schema=daily_schema, header=False)


### Checking with daily 

In [16]:

# Get distinct station IDs and rename the column
daily_ids = daily_2025_df.select("ID").distinct().withColumnRenamed("ID", "DAILY_ID")

# Count them
unique_station_count = daily_ids.count()
print(f"Number of unique stations in 2025 daily data: {unique_station_count}")


[Stage 11:>                                                         (0 + 1) / 1]

Number of unique stations in 2025 daily data: 32224


In [17]:

# Select distinct station IDs where LASTYEAR is 2025
station_ids = enriched_stations.filter(col("LASTYEAR") == 2025) \
    .select("ID").distinct().withColumnRenamed("ID", "STATION_ID")

# Count and print
station_count = station_ids.count()
print(f"Number of unique stations in 2025 enriched_stations data: {station_count}")


Number of unique stations in 2025 enriched_stations data: 30735


In [18]:

# Daily not in stations
extra_in_daily = daily_ids.join(station_ids, daily_ids.DAILY_ID == station_ids.STATION_ID, "left_anti")
print("Extra in daily:", extra_in_daily.count())


[Stage 26:>                                                         (0 + 1) / 1]

Extra in daily: 1489


**Note:** There are 1489 station IDs in the daily dataset for 2025 that do not updated in the station metadata last year. 


In [19]:

extra_station_info = extra_in_daily.join(
    enriched_stations,
    extra_in_daily.DAILY_ID == enriched_stations.ID,
    "left"
)


In [20]:

# Group by country and collect station IDs
country_station_ids = extra_station_info.groupBy("COUNTRY_NAME") \
    .agg(
        count("*").alias("station_count"),
        collect_list("ID").alias("station_ids")
    ) \
    .orderBy("station_count", ascending=False)


# Show only first 5 station IDs for each country
limited_station_ids = country_station_ids.withColumn("station_ids", slice("station_ids", 1, 5)) 
limited_station_ids.show(truncate=False)
# Here I slice the result


[Stage 36:>                                                         (0 + 1) / 1]

+----------------------------+-------------+-----------------------------------------------------------------+
|COUNTRY_NAME                |station_count|station_ids                                                      |
+----------------------------+-------------+-----------------------------------------------------------------+
|United States               |1475         |[USR0000TPRE, USR0000WDIR, USR0000TCHU, USR0000IPIE, USR0000MBDA]|
|Canada                      |4            |[CA1QC000076, CA008501100, CA1MB000346, CA1NS000142]             |
|Ethiopia                    |3            |[ET000063333, ET000063403, ET000063402]                          |
|Puerto Rico [United States] |2            |[RQC00665097, RQ1PRVQ0003]                                       |
|Russia                      |2            |[RSM00029645, RSM00032583]                                       |
|Paraguay                    |2            |[PAM00086170, PA000086086]                                       |
|


- There are 30,735 stations listed as active in the station metadata for 2025.
- But in the 2025 daily climate data, there are 32,224 unique stations. 
- This means 1,489 stations have daily data but are not marked as active in the station metadata.

## (a.3) How many stations are in each of the GSN, HCN and CRN?

In [21]:

enriched_stations.show(5)


+-----------+----------+------------+--------+---------+---------+--------------------+--------+------------+------+--------------------+----------+---------+--------+--------------+-------------+--------------+
|         ID|STATE_CODE|COUNTRY_CODE|LATITUDE|LONGITUDE|ELEVATION|                NAME|GSN_FLAG|HCN_CRN_FLAG|WMO_ID|        COUNTRY_NAME|STATE_NAME|FIRSTYEAR|LASTYEAR|TOTAL_ELEMENTS|CORE_ELEMENTS|OTHER_ELEMENTS|
+-----------+----------+------------+--------+---------+---------+--------------------+--------+------------+------+--------------------+----------+---------+--------+--------------+-------------+--------------+
|AE000041196|      NULL|          AE|  25.333|   55.517|     34.0|SHARJAH INTER. AI...|     GSN|        NULL| 41196|United Arab Emira...|      NULL|     1944|    2025|             4|            3|             1|
|AEM00041218|      NULL|          AE|  24.262|   55.609|    264.9|AL AIN INTL      ...|    NULL|        NULL| 41218|United Arab Emira...|      NULL|    

In [22]:

# Count number of stations in each network
gsn_count = enriched_stations.filter(col("GSN_FLAG").isNotNull()).count()
hcn_count = enriched_stations.filter(col("HCN_CRN_FLAG") == "HCN").count()
crn_count = enriched_stations.filter(col("HCN_CRN_FLAG") == "CRN").count()

print(f"Number of stations in GSN: {gsn_count}")
print(f"Number of stations in HCN: {hcn_count}")
print(f"Number of stations in CRN: {crn_count}")


Number of stations in GSN: 991
Number of stations in HCN: 1218
Number of stations in CRN: 234


 #### Checking

In [23]:

# Count GSN_FLAG values
enriched_stations.groupBy("GSN_FLAG").count().show()

# Count HCN_CRN_FLAG values
enriched_stations.groupBy("HCN_CRN_FLAG").count().show()


+--------+------+
|GSN_FLAG| count|
+--------+------+
|     GSN|   991|
|    NULL|128666|
+--------+------+

+------------+------+
|HCN_CRN_FLAG| count|
+------------+------+
|         HCN|  1218|
|        NULL|128205|
|         CRN|   234|
+------------+------+



In [24]:

total_stations = enriched_stations.select("ID").distinct().count()
print(f"Total number of unique stations in metadata: {total_stations}")


Total number of unique stations in metadata: 129657





**GSN_FLAG:**
  
- GSN: 991 stations

- NULL: 128,666 stations

- Total = 991 + 128,666 = 129,657


**HCN_CRN_FLAG:**
  
- HCN: 1,218 stations

- CRN: 234 stations

- NULL: 128,205 stations
  
- Total = 1,218 + 234 + 128,205 = 129,657


## (a.4) Are there any stations that are in more than one of these networks?

In [25]:

network_stations = enriched_stations.filter(
    (col("GSN_FLAG").isNotNull()) | (col("HCN_CRN_FLAG").isNotNull())
)

network_counts = (
    network_stations
    .groupby("GSN_FLAG", "HCN_CRN_FLAG")
    .agg(count("*").alias("count"))
    .orderBy("GSN_FLAG", "HCN_CRN_FLAG")
)

network_counts.show()


+--------+------------+-----+
|GSN_FLAG|HCN_CRN_FLAG|count|
+--------+------------+-----+
|    NULL|         CRN|  234|
|    NULL|         HCN| 1203|
|     GSN|        NULL|  976|
|     GSN|         HCN|   15|
+--------+------------+-----+



## (b.1) How many stations are there in the Southern Hemisphere?

In [26]:

enriched_stations.selectExpr("min(LATITUDE)", "max(LATITUDE)").show()


+-------------+-------------+
|min(LATITUDE)|max(LATITUDE)|
+-------------+-------------+
|        -90.0|        83.65|
+-------------+-------------+



In [27]:

enriched_stations.filter(col("LATITUDE") < 0).select("ID", "LATITUDE", "COUNTRY_NAME").show(10, truncate=False)


+-----------+--------+-------------------------------+
|ID         |LATITUDE|COUNTRY_NAME                   |
+-----------+--------+-------------------------------+
|AO000066270|-11.417 |Angola                         |
|AQC00914594|-14.3333|American Samoa [United States] |
|AQC00914873|-14.35  |American Samoa [United States] |
|AQW00061705|-14.3306|American Samoa [United States] |
|AR000000002|-29.82  |Argentina                      |
|AR000087078|-24.7   |Argentina                      |
|AR000087925|-51.617 |Argentina                      |
|AR000875850|-34.583 |Argentina                      |
|ARM00087016|-23.153 |Argentina                      |
|ARM00087022|-22.62  |Argentina                      |
+-----------+--------+-------------------------------+
only showing top 10 rows



In [28]:

southern_count = enriched_stations.filter(col("LATITUDE") < 0).select("ID").distinct().count()
print(f"Number of stations in the Southern Hemisphere: {southern_count}")


Number of stations in the Southern Hemisphere: 25357


In [29]:

northern_count = enriched_stations.filter(col("LATITUDE") >= 0).select("ID").distinct().count()
print(f"Number of stations in the Northern Hemisphere: {northern_count}")


Number of stations in the Northern Hemisphere: 104300



#### Note:

- Only about 19.6% of all stations are in the Southern Hemisphere
- About 80.4% are in the Northern Hemisphere


## (b.2) How many stations are there in total in the territories of the United States,  (~United States)?

In [30]:


# Filter for U.S. territories (exclude mainland United States)
us_territory_stations = enriched_stations.filter(
    (trim(col("COUNTRY_NAME")).contains("United States")) &
    (trim(col("COUNTRY_NAME")) != "United States")
)


In [31]:

# Count stations in U.S. territories
us_territory_count = us_territory_stations.count()

us_territory_count


414

In [32]:

# Group by country name (i.e., each U.S. territory) and count the number of stations
territory_counts = us_territory_stations.groupBy("COUNTRY_NAME").count().orderBy("count", ascending=False)
territory_counts.show(truncate=False)

+-----------------------------------------+-----+
|COUNTRY_NAME                             |count|
+-----------------------------------------+-----+
|Puerto Rico [United States]              |260  |
|Virgin Islands [United States]           |77   |
|Guam [United States]                     |34   |
|American Samoa [United States]           |21   |
|Northern Mariana Islands [United States] |11   |
|Johnston Atoll [United States]           |4    |
|Palmyra Atoll [United States]            |3    |
|Midway Islands [United States}           |3    |
|Wake Island [United States]              |1    |
+-----------------------------------------+-----+



In [33]:

# Print the result
print(f"Total number of stations in U.S. territories (excluding mainland U.S.): {us_territory_count}")


Total number of stations in U.S. territories (excluding mainland U.S.): 414


### Checking 

In [34]:


mainland_us = enriched_stations.filter(trim(col("COUNTRY_NAME")) == "United States")
mainland_count = mainland_us.select("ID").distinct().count()
print(f"Mainland U.S. station count: {mainland_count}")


Mainland U.S. station count: 75846


In [35]:

all_us_related = enriched_stations.filter(col("COUNTRY_NAME").contains("United States"))
all_us_related_count = all_us_related.select("ID").distinct().count()


In [36]:

print("Territories:", us_territory_count)
print("Mainland US:", mainland_count)
print("Combined:", us_territory_count + mainland_count)
print("All US-related:", all_us_related_count)


Territories: 414
Mainland US: 75846
Combined: 76260
All US-related: 76260


## (c.1) Count total number of stations in each country, and join these counts onto countries and save 

In [37]:

# Count stations per country with country code
stations_per_country = (
    enriched_stations
    .groupBy("COUNTRY_NAME", "COUNTRY_CODE")
    .agg(count("ID").alias("STATION_COUNT"))
    .orderBy("STATION_COUNT", ascending=False)
)

# Show top 10
print("Top 10 countries by station count:")
stations_per_country.show(10, truncate=False)


Top 10 countries by station count:
+--------------+------------+-------------+
|COUNTRY_NAME  |COUNTRY_CODE|STATION_COUNT|
+--------------+------------+-------------+
|United States |US          |75846        |
|Australia     |AS          |17088        |
|Canada        |CA          |9269         |
|Brazil        |BR          |5989         |
|Mexico        |MX          |5249         |
|India         |IN          |3807         |
|Sweden        |SW          |1721         |
|South Africa  |SF          |1166         |
|Germany       |GM          |1123         |
|Russia        |RS          |1123         |
+--------------+------------+-------------+
only showing top 10 rows



In [38]:

# Show bottom 10
print("Bottom 10 countries by station count:")
stations_per_country.orderBy("STATION_COUNT").show(10, truncate=False)


Bottom 10 countries by station count:
+------------------------------------+------------+-------------+
|COUNTRY_NAME                        |COUNTRY_CODE|STATION_COUNT|
+------------------------------------+------------+-------------+
|Cayman Islands [United Kingdom]     |CJ          |1            |
|Iraq                                |IZ          |1            |
|Cambodia                            |CB          |1            |
|Malta                               |MT          |1            |
|Saint Pierre and Miquelon [France]  |SB          |1            |
|Cocos (Keeling) Islands [Australia] |CK          |1            |
|Maldives                            |MV          |1            |
|Macau S.A.R                         |MC          |1            |
|Jan Mayen [Norway]                  |JN          |1            |
|Tokelau [New Zealand]               |TL          |1            |
+------------------------------------+------------+-------------+
only showing top 10 rows



In [39]:

# Count how many countries have exactly one station
one_station_countries = stations_per_country.filter("STATION_COUNT = 1").count()

print(f"Number of countries with exactly one station: {one_station_countries}")


Number of countries with exactly one station: 41


In [40]:

countries = spark.read.parquet("wasbs://campus-user@madsstorage002.blob.core.windows.net/asa349/countries.parquet")


In [41]:
countries.show(5, truncate=False)

+------------+---------------------+
|COUNTRY_CODE|COUNTRY_NAME         |
+------------+---------------------+
|AC          |Antigua and Barbuda  |
|AE          |United Arab Emirates |
|AF          |Afghanistan          |
|AG          |Algeria              |
|AJ          |Azerbaijan           |
+------------+---------------------+
only showing top 5 rows



In [42]:

# Join with countries using on COUNTRY_NAME (from countries)
countries_with_counts = (
    countries.alias("c")
    .join(stations_per_country.alias("s"), on="COUNTRY_CODE", how="left")
    .select(
        col("c.COUNTRY_CODE"),
        col("c.COUNTRY_NAME"),
        col("s.STATION_COUNT")
    )
)

In [43]:
countries_with_counts.show(10, truncate=False)

+------------+-------------------------------+-------------+
|COUNTRY_CODE|COUNTRY_NAME                   |STATION_COUNT|
+------------+-------------------------------+-------------+
|AC          |Antigua and Barbuda            |2            |
|AE          |United Arab Emirates           |4            |
|AF          |Afghanistan                    |4            |
|AG          |Algeria                        |87           |
|AJ          |Azerbaijan                     |66           |
|AL          |Albania                        |3            |
|AM          |Armenia                        |53           |
|AO          |Angola                         |6            |
|AQ          |American Samoa [United States] |21           |
|AR          |Argentina                      |101          |
+------------+-------------------------------+-------------+
only showing top 10 rows



In [44]:

# Save countries_with_counts DataFrame as Parquet
countries_with_counts.write.mode("overwrite").parquet(
    "wasbs://campus-user@madsstorage002.blob.core.windows.net/asa349/countries_with_counts.parquet"
)


25/04/09 12:32:07 WARN AzureFileSystemThreadPoolExecutor: Disabling threads for Delete operation as thread count 0 is <= 1
25/04/09 12:32:08 WARN AzureFileSystemThreadPoolExecutor: Disabling threads for Delete operation as thread count 0 is <= 1


## (c.2) Count the total number of stations in each states, and join these counts onto states ans save

In [45]:

# Filter out NULL state names
valid_states = enriched_stations.filter(col("STATE_NAME").isNotNull())

# Count stations per valid state
state_counts = valid_states.groupBy("STATE_NAME", "STATE_CODE") \
                           .count() \
                           .withColumnRenamed("count", "STATION_COUNT") \
                           .orderBy(col("STATION_COUNT").desc())

# Show top 10 states
print("Top 10 states by station count:")
state_counts.show(10, truncate=False)


Top 10 states by station count:
+--------------+----------+-------------+
|STATE_NAME    |STATE_CODE|STATION_COUNT|
+--------------+----------+-------------+
|TEXAS         |TX        |6472         |
|COLORADO      |CO        |4784         |
|CALIFORNIA    |CA        |3166         |
|NORTH CAROLINA|NC        |2747         |
|MINNESOTA     |MN        |2675         |
|NEBRASKA      |NE        |2436         |
|KANSAS        |KS        |2401         |
|NEW MEXICO    |NM        |2295         |
|FLORIDA       |FL        |2244         |
|ILLINOIS      |IL        |2234         |
+--------------+----------+-------------+
only showing top 10 rows



In [46]:

# Show bottom 10 states
print("Bottom 10 states by station count:")
state_counts.orderBy(col("STATION_COUNT").asc()).show(10, truncate=False)


Bottom 10 states by station count:
+---------------------------+----------+-------------+
|STATE_NAME                 |STATE_CODE|STATION_COUNT|
+---------------------------+----------+-------------+
|NORTHERN MARIANA ISLANDS   |MP        |11           |
|U.S. MINOR OUTLYING ISLANDS|UM        |11           |
|PALAU                      |PW        |13           |
|DISTRICT OF COLUMBIA       |DC        |18           |
|MARSHALL ISLANDS           |MH        |21           |
|AMERICAN SAMOA             |AS        |21           |
|GUAM                       |GU        |34           |
|MICRONESIA                 |FM        |38           |
|VIRGIN ISLANDS             |VI        |77           |
|PRINCE EDWARD ISLAND       |PE        |94           |
+---------------------------+----------+-------------+
only showing top 10 rows



In [47]:

states = spark.read.parquet("wasbs://campus-user@madsstorage002.blob.core.windows.net/asa349/states.parquet")


In [48]:
states.show(5, truncate=False)

+----------+-----------------------------------------------+
|STATE_CODE|STATE_NAME                                     |
+----------+-----------------------------------------------+
|AB        |ALBERTA                                        |
|AK        |ALASKA                                         |
|AL        |ALABAMA                                        |
|AR        |ARKANSAS                                       |
|AS        |AMERICAN SAMOA                                 |
+----------+-----------------------------------------------+
only showing top 5 rows



In [49]:

# Join states table with state_counts to include station counts
states_with_counts = (
    states.alias("s")
    .join(state_counts.alias("c"), on="STATE_CODE", how="left")
    .select(
        col("s.STATE_CODE"),
        col("s.STATE_NAME"),
        col("c.STATION_COUNT")
    )
)

In [50]:

states_with_counts.write.mode("overwrite").parquet("wasbs://campus-user@madsstorage002.blob.core.windows.net/asa349/states_with_counts.parquet")


25/04/09 12:32:21 WARN AzureFileSystemThreadPoolExecutor: Disabling threads for Delete operation as thread count 0 is <= 1
25/04/09 12:32:32 WARN AzureFileSystemThreadPoolExecutor: Disabling threads for Delete operation as thread count 0 is <= 1



# Question-2


## (a.1) Spark Function for Distance Calculation

In [51]:

# Haversine formula to compute distance in kilometers
def haversine(lat1, lon1, lat2, lon2):
    R = 6371  # Radius of the Earth in km
    phi1, phi2 = math.radians(lat1), math.radians(lat2)
    delta_phi = math.radians(lat2 - lat1)
    delta_lambda = math.radians(lon2 - lon1)
    
    a = math.sin(delta_phi / 2.0) ** 2 + math.cos(phi1) * math.cos(phi2) * math.sin(delta_lambda / 2.0) ** 2
    c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a))
    
    return R * c

# Convert function to Spark UDF
haversine_udf = udf(haversine, DoubleType())


## (a.2) Test this function by using CROSS JOIN on a small subset of stations

In [52]:

# Select a small subset of stations with necessary columns
stations_subset = enriched_stations.select("ID", "LATITUDE", "LONGITUDE", "COUNTRY_NAME").limit(10)

# Perform a CROSS JOIN to compare all pairs of stations
cross_joined_stations = stations_subset.alias("s1").crossJoin(stations_subset.alias("s2"))

# Compute distance for each pair of stations
distance_df = cross_joined_stations.withColumn(
    "distance_km",
    haversine_udf(
        col("s1.LATITUDE"), col("s1.LONGITUDE"),
        col("s2.LATITUDE"), col("s2.LONGITUDE")
    )
)

# Show results with country names included
distance_df.select(
    col("s1.ID").alias("STATION1_ID"),
    col("s1.COUNTRY_NAME").alias("STATION1_COUNTRY"),
    col("s2.ID").alias("STATION2_ID"),
    col("s2.COUNTRY_NAME").alias("STATION2_COUNTRY"),
    col("DISTANCE_KM")
).show(5, truncate=False)



+-----------+---------------------+-----------+---------------------+------------------+
|STATION1_ID|STATION1_COUNTRY     |STATION2_ID|STATION2_COUNTRY     |DISTANCE_KM       |
+-----------+---------------------+-----------+---------------------+------------------+
|AE000041196|United Arab Emirates |AE000041196|United Arab Emirates |0.0               |
|AE000041196|United Arab Emirates |AEM00041218|United Arab Emirates |119.45143040410233|
|AE000041196|United Arab Emirates |AGE00147715|Algeria              |4637.495558536303 |
|AE000041196|United Arab Emirates |AGE00147717|Algeria              |5317.367078902586 |
|AE000041196|United Arab Emirates |AGE00147719|Algeria              |5121.417908402876 |
+-----------+---------------------+-----------+---------------------+------------------+
only showing top 5 rows



## (b.1) Computing Pairwise Distances in New Zealand

In [53]:

# Filter only stations in "New Zealand"
nz_stations = enriched_stations.filter(trim(col("COUNTRY_NAME")) == "New Zealand") \
                               .select("ID", "LATITUDE", "LONGITUDE", "NAME")

# Show a few rows to verify
nz_stations.show(20, truncate=False)


+-----------+--------+---------+------------------------------+
|ID         |LATITUDE|LONGITUDE|NAME                          |
+-----------+--------+---------+------------------------------+
|NZ000936150|-42.717 |170.983  |HOKITIKA AERODROME            |
|NZ000939450|-52.55  |169.167  |CAMPBELL ISLAND AWS           |
|NZM00093439|-41.333 |174.8    |WELLINGTON AERO AWS           |
|NZM00093929|-50.483 |166.3    |ENDERBY ISLAND AWS            |
|NZ000093417|-40.9   |174.983  |PARAPARAUMU AWS               |
|NZ000093844|-46.417 |168.333  |INVERCARGILL AIRPOR           |
|NZ000937470|-44.517 |169.9    |TARA HILLS                    |
|NZ000939870|-43.95  |-176.567 |CHATHAM ISLANDS AWS           |
|NZM00093110|-37.0   |174.8    |AUCKLAND AERO AWS             |
|NZM00093781|-43.489 |172.532  |CHRISTCHURCH INTL             |
|NZ000093994|-29.25  |-177.917 |RAOUL ISL/KERMADEC            |
|NZ000933090|-39.017 |174.183  |NEW PLYMOUTH AWS              |
|NZ000093012|-35.1   |173.267  |KAITAIA 

In [54]:

# Verification
enriched_stations.select("COUNTRY_NAME").distinct().filter(col("COUNTRY_NAME").like("%Zealand%")).show(100, False)


+--------------------------+
|COUNTRY_NAME              |
+--------------------------+
|New Zealand               |
|Tokelau [New Zealand]     |
|Niue [New Zealand]        |
|Cook Islands [New Zealand]|
+--------------------------+



In [55]:

# Perform CROSS JOIN for pairwise comparisons
nz_pairs = nz_stations.alias("s1").crossJoin(nz_stations.alias("s2"))


In [56]:

# Compute distances using Haversine UDF
nz_distance_df = nz_pairs.withColumn(
    "distance_km",
    haversine_udf(
        col("s1.LATITUDE"), col("s1.LONGITUDE"),
        col("s2.LATITUDE"), col("s2.LONGITUDE")
    )
)

In [57]:

# Select required columns for output with station names
nz_distance_df = nz_distance_df.select(
    col("s1.ID").alias("STATION1_ID"),
    col("s1.NAME").alias("STATION1_NAME"),
    col("s2.ID").alias("STATION2_ID"),
    col("s2.NAME").alias("STATION2_NAME"),
    col("DISTANCE_KM")
)


In [58]:
# Show sample output
nz_distance_df.show(5, truncate=False)

+-----------+------------------------------+-----------+------------------------------+------------------+
|STATION1_ID|STATION1_NAME                 |STATION2_ID|STATION2_NAME                 |DISTANCE_KM       |
+-----------+------------------------------+-----------+------------------------------+------------------+
|NZ000936150|HOKITIKA AERODROME            |NZ000936150|HOKITIKA AERODROME            |0.0               |
|NZ000936150|HOKITIKA AERODROME            |NZ000939450|CAMPBELL ISLAND AWS           |1101.719032102152 |
|NZ000936150|HOKITIKA AERODROME            |NZM00093439|WELLINGTON AERO AWS           |350.79606843225486|
|NZ000936150|HOKITIKA AERODROME            |NZM00093929|ENDERBY ISLAND AWS            |934.2477752798079 |
|NZ000936150|HOKITIKA AERODROME            |NZ000093417|PARAPARAUMU AWS               |388.1759742441008 |
+-----------+------------------------------+-----------+------------------------------+------------------+
only showing top 5 rows



In [59]:

# Define output path
output_relative_path_nz = f"{username}/nz_station_distances"
output_path_nz = f"wasbs://{azure_user_container_name}@{azure_account_name}.blob.core.windows.net/{output_relative_path_nz}"


In [60]:

#  Save to output directory
nz_distance_df.write.mode("overwrite").parquet(output_path_nz)


25/04/09 12:32:58 WARN AzureFileSystemThreadPoolExecutor: Disabling threads for Delete operation as thread count 0 is <= 1
25/04/09 12:33:01 WARN AzureFileSystemThreadPoolExecutor: Disabling threads for Delete operation as thread count 0 is <= 1


## (b.2) What two stations are geographically closest together in New Zealand?

In [61]:

# Filter out same-station comparisons (where distance = 0)
closest_pair = nz_distance_df.filter(col("distance_km") > 0) \
                             .orderBy(col("distance_km").asc()) \
                             .limit(1)

closest_pair.show(truncate=False)


25/04/09 12:33:10 WARN ExtractPythonUDFFromJoinCondition: The join condition:(haversine(LATITUDE#3, LONGITUDE#4, LATITUDE#925, LONGITUDE#926)#948 > 0.0) of the join plan contains PythonUDF only, it will be moved out and the join plan will be turned to cross join.


+-----------+------------------------------+-----------+------------------------------+-----------------+
|STATION1_ID|STATION1_NAME                 |STATION2_ID|STATION2_NAME                 |DISTANCE_KM      |
+-----------+------------------------------+-----------+------------------------------+-----------------+
|NZ000093417|PARAPARAUMU AWS               |NZM00093439|WELLINGTON AERO AWS           |50.52885002286977|
+-----------+------------------------------+-----------+------------------------------+-----------------+



In [62]:

# Filter for the two station IDs
closest_stations = enriched_stations.filter(
    col("ID").isin("NZ000093417", "NZM00093439")
)

# Select columns to display
closest_stations.select("ID", "NAME", "LATITUDE", "LONGITUDE").show(truncate=False)


+-----------+------------------------------+--------+---------+
|ID         |NAME                          |LATITUDE|LONGITUDE|
+-----------+------------------------------+--------+---------+
|NZM00093439|WELLINGTON AERO AWS           |-41.333 |174.8    |
|NZ000093417|PARAPARAUMU AWS               |-40.9   |174.983  |
+-----------+------------------------------+--------+---------+



# Question 3

In [63]:

daily_schema = StructType([
    StructField("ID", StringType(), False),         
    StructField("DATE", StringType(), False),       
    StructField("ELEMENT", StringType(), False),    
    StructField("VALUE", FloatType(), False),       
    StructField("MEASUREMENT_FLAG", StringType(), True),  
    StructField("QUALITY_FLAG", StringType(), True),      
    StructField("SOURCE_FLAG", StringType(), True),       
    StructField("OBSERVATION_TIME", StringType(), True)   
])


In [64]:

daily_path = "wasbs://campus-data@madsstorage002.blob.core.windows.net/ghcnd/daily/*.csv.gz"
daily_df = spark.read.csv(daily_path, schema=daily_schema, header=False)

# Convert DATE to proper format
daily = daily_df.withColumn("DATE", to_date(col("DATE"), "yyyyMMdd"))

# Show sample data
daily.show(5)


+-----------+----------+-------+-----+----------------+------------+-----------+----------------+
|         ID|      DATE|ELEMENT|VALUE|MEASUREMENT_FLAG|QUALITY_FLAG|SOURCE_FLAG|OBSERVATION_TIME|
+-----------+----------+-------+-----+----------------+------------+-----------+----------------+
|ASN00037091|2010-01-01|   PRCP|  0.0|            NULL|        NULL|          a|            NULL|
|ASN00037104|2010-01-01|   PRCP| 86.0|            NULL|        NULL|          a|            NULL|
|ASN00037107|2010-01-01|   PRCP|  0.0|            NULL|        NULL|          a|            NULL|
|ASN00037112|2010-01-01|   PRCP| 50.0|            NULL|        NULL|          a|            NULL|
|ASN00037119|2010-01-01|   PRCP|  0.0|            NULL|        NULL|          a|            NULL|
+-----------+----------+-------+-----+----------------+------------+-----------+----------------+
only showing top 5 rows



## (a.1) Count the number of rows in daily.


In [65]:

# Count the number of rows
num_rows = daily.count()

# Print the result
print(f"Total number of rows in daily dataset: {num_rows}")


[Stage 154:====================================================>(105 + 1) / 106]

Total number of rows in daily dataset: 3139143397


## (b.1) Filter daily using the filter

In [66]:

# Define the core elements
core_elements = ["TMAX", "TMIN", "PRCP", "SNOW", "SNWD"]

# Filter for only the five core elements
filtered_daily = daily.filter(col("ELEMENT").isin(core_elements))


In [67]:

output_path = "wasbs://campus-user@madsstorage002.blob.core.windows.net/asa349/filtered_daily_core.parquet"
filtered_daily.write.mode("overwrite").parquet(output_path)


25/04/09 12:35:29 WARN AzureFileSystemThreadPoolExecutor: Disabling threads for Delete operation as thread count 0 is <= 1
25/04/09 12:47:29 WARN AzureFileSystemThreadPoolExecutor: Disabling threads for Delete operation as thread count 0 is <= 1


## (b.2) How many observations are there for each of the five core elements?

In [68]:

# Input path
input_path = "wasbs://campus-user@madsstorage002.blob.core.windows.net/asa349/filtered_daily_core.parquet"

# Read the Parquet file
filtered_daily = spark.read.parquet(input_path)

# Show a few rows 
filtered_daily.show(5)


[Stage 159:>                                                        (0 + 1) / 1]

+-----------+----------+-------+-----+----------------+------------+-----------+----------------+
|         ID|      DATE|ELEMENT|VALUE|MEASUREMENT_FLAG|QUALITY_FLAG|SOURCE_FLAG|OBSERVATION_TIME|
+-----------+----------+-------+-----+----------------+------------+-----------+----------------+
|ASN00037091|2010-01-01|   PRCP|  0.0|            NULL|        NULL|          a|            NULL|
|ASN00037104|2010-01-01|   PRCP| 86.0|            NULL|        NULL|          a|            NULL|
|ASN00037107|2010-01-01|   PRCP|  0.0|            NULL|        NULL|          a|            NULL|
|ASN00037112|2010-01-01|   PRCP| 50.0|            NULL|        NULL|          a|            NULL|
|ASN00037119|2010-01-01|   PRCP|  0.0|            NULL|        NULL|          a|            NULL|
+-----------+----------+-------+-----+----------------+------------+-----------+----------------+
only showing top 5 rows



In [69]:

# Count the occurrences of each core element
element_counts = filtered_daily.groupBy("ELEMENT").agg(count("*").alias("count"))

# Show the results
element_counts.show()


[Stage 160:======================================================>(95 + 1) / 96]

+-------+----------+
|ELEMENT|     count|
+-------+----------+
|   SNOW| 359249644|
|   TMIN| 458928768|
|   SNWD| 300711620|
|   PRCP|1079767077|
|   TMAX| 460114659|
+-------+----------+



## (b.3) Which element has the most observations?

In [70]:

# Find the element with the most observations
most_frequent_element = element_counts.orderBy(col("count").desc()).first()

# Print the result
print(f"The element with the most observations is {most_frequent_element['ELEMENT']} with {most_frequent_element['count']} records.")


[Stage 163:======================================================>(95 + 1) / 96]

The element with the most observations is PRCP with 1079767077 records.



## (c.1) How many observations of TMAX do not have a corresponding observation of TMIN.

### Using Collect Set

In [71]:

# Group by ID and DATE, collect observed elements (TMAX, TMIN, etc.)
grouped = filtered_daily.groupBy("ID", "DATE") \
    .agg(collect_set("ELEMENT").alias("elements"))


In [72]:

# Filter have TMAX but not TMIN
tmax_only = grouped.filter(array_contains("elements", "TMAX")) \
                   .filter(~array_contains("elements", "TMIN"))


In [73]:

# Count total unmatched TMAX observations
print("Unmatched TMAX count:", tmax_only.count())


[Stage 168:====================================================>  (46 + 2) / 48]

Unmatched TMAX count: 10660214


## (c.2) How many unique stations contributed to these observations?

In [74]:

# Count unique stations contributing these observations
print("Unique station count:", tmax_only.select("ID").distinct().count())


[Stage 174:=====================================================> (47 + 1) / 48]

Unique station count: 28754


### Checked withLeft anti Join

In [84]:

# Filter TMAX and TMIN observations
tmax_df = filtered_daily.filter(col("ELEMENT") == "TMAX").select("ID", "DATE", "ELEMENT").distinct()
tmin_df = filtered_daily.filter(col("ELEMENT") == "TMIN").select("ID", "DATE", "ELEMENT").distinct()


In [87]:

# Left anti join 
tmax_without_tmin = tmax_df.join(tmin_df, on=["ID", "DATE"], how="left_anti")


In [88]:

# Count the number of unmatched TMAX observations
unmatched_tmax_count = tmax_without_tmin.count()


In [89]:

#  Count unique stations 
unique_station_count = tmax_without_tmin.select("ID").distinct().count()


In [91]:

# Show the results
print(f"Number of TMAX observations without TMIN: {unmatched_tmax_count}")


Number of TMAX observations without TMIN: 10660214


In [90]:

print(f"Number of unique stations with unmatched TMAX: {unique_station_count}")


Number of unique stations with unmatched TMAX: 28754


In [96]:
# Run this cell before closing the notebook or kill your spark application by hand using the link in the Spark UI

stop_spark()

25/04/07 22:19:04 WARN ExecutorPodsWatchSnapshotSource: Kubernetes client has been closed.
